### ISSUE TITLE
Add vips driver

### ISSUE THREAD

**Issue by anshu8858** [Intervention/image#1119]
Please add **`vips`** for driver. It takes less RAM.
[Stack Overflow](https://stackoverflow.com/questions/63671182/resize-the-image-using-php-vips-library)
[PECL](https://pecl.php.net/package/vips)

**Comment by mstaack** [comment]
try this, works well for me:

https://github.com/osiemsiedem/intervention-image-vips-driver

**Comment by jcupitt** [comment]
Hello, I had a quick look at that driver and made an issue with a couple of things I spotted: https://github.com/osiemsiedem/intervention-image-vips-driver/issues/27

might be useful.

**Comment by mstaack** [comment]
@jcupitt oh great, thanks

**Comment by ARehmanMahi** [comment]
I would advise against the `osiemsiedem` driver. The maintainer doesn't cooperate with `intervention` devs & doesn't even reply to the other maintainers of the vips libs itself `@jcupitt`.  Thinks intervention is dead yet tries to have an adapter for the "dead" lib.

There have been talks of such adapters for 5 some years. Yet no progress on it. Seems like a lost cause.
I'm just a junior php dev, please don't come at me with PRs lol.

Intervention & vips are both amazing projects. Would love to have them working together.

**Comment by jcupitt** [comment]
The osiemsiedem driver is MIT licenced, so it could be forked and maintained by intervention, if necessary. I'm not offering to do the work myself, of course ahem

**Comment by ARehmanMahi** [comment]
> The osiemsiedem driver is MIT licenced, so it could be forked and maintained by intervention, if necessary. I'm not offering to do the work myself, of course ahem

That is why I'm going to remove Intervention and directly use vips-php, not that difficult to use standalone.

**Comment by olivervogel** [comment]
> That is why I'm going to remove Intervention and directly use vips-php, not that difficult to use standalone.

I agree. Intervention Image is basically just a level of comfort and convenience. If one intend to use only VIPS (or any other layer) anyway, there is no reason not to this the library directly.

**Comment by GuySartorelli** [comment]
https://github.com/osiemsiedem/intervention-image-vips-driver is outdated now - hasn't been updated for the v3 rewrite.

I'm interested in trying my hand at adding vips support natively to intervention/image - but that's obviously a massive undertaking and I'll need a fair amount of support and review to get going.

@olivervogel If I were to tackle this, what would be your preferred approach?
Would you be open to having a `vips` branch, and I can target incremental PRs to that branch to slowly build up the feature?
Would you prefer I open a PR targetting the 'develop' branch, and have it in draft with incremental commits in that same PR?
Would you rather I not open a PR until the feature is 100% complete?

There is of course always the option of saying "intervention/image will not have vips support" which is a valid stance to take since it will add a lot of additional code to maintain.

**Comment by olivervogel** [comment]
I have often thought about integrating Vips as a third driver and love the idea to have such a tool.

This driver would however definitely create a new dependency with [php-vips](https://github.com/libvips/php-vips). For this reason I don't think it's a good idea to integrate it directly into `intervention/image`. I want to prevent the Vips PHP Binding Composer Package `jcupitt/vips` from being always automatically installed, even if the user does not intend to use Vips.

In my opinion, a better solution is to build the Vips driver as another package like `intervention/image-driver-vips` for example. This could then be installed optionally and potentially unused dependencies could be avoided.

However, I haven't made much progress with the implementation yet. Everything is still in a very early stage and I would call it a case study to see how I can best integrate vips. There are no tests yet either.

I'm still not sure if all the features I've implemented for GD and Imagick will work for Vips. A first goal might be to implement at least the image resizing functions for a Vips driver and writing tests for the features implemented so far.

I put the [current state into a public repository today](https://github.com/Intervention/image-driver-vips), but it is not yet listed on Packagist. The driver is able to read (animated) images, do simple image resizing with `resize()` and encode in Jpeg format.

If you want, you can take the repository and continue working on it. Since this is a project independent of the main repository and the progress is still very early and WIP, any small changes and also larger features can be submitted as PR in the external vips driver repository (https://github.com/Intervention/image-driver-vips). I do not expect anything to be complete and fully functional yet.

**Comment by jcupitt** [comment]
The upcoming libvips 8.16 has a new chapter in the docs for a developer checklist:

https://github.com/libvips/libvips/blob/master/doc/Developer-checklist.md

It tries to give some general guidelines for designing things that use libvips. It might be helpful.

**Comment by binaryfire** [comment]
Hi @olivervogel. Thanks for all your hard work on Invervention. 

I noticed you've made a few recent commits to the [vips extension](https://github.com/Intervention/image-driver-vips). I'm about to begin work on a high traffic image-focused PHP app and deciding between using php-vips directly or sticking with Intervention. In July you mentioned it was an experiment- has that changed and do you plan to release it as a production-ready extension?

**Comment by olivervogel** [comment]
> has that changed and do you plan to release it as a production-ready extension?

Unfortunately not. There is currently no schedule or date.

**Comment by olivervogel** [comment]
@deluxetom and I have been working together for the last few weeks on the official [libvips driver for Intervention Image](https://github.com/Intervention/image-driver-vips).

https://github.com/Intervention/image-driver-vips

There is now a official beta version which supports all features and only has two minor limitations:

- Due to the technical characteristics of libvips, it is currently not possible to implement color quantization via `ImageInterface::reduceColors()` as intended. However, there is a [pull request in libvips](https://github.com/libvips/php-vips/issues/256#issuecomment-2575872401) that enables this feature and it may be integrated here the future as well.

- With PHP on macOS, font files are not recognized in the `ImageInterface::text()` call because Quartz as a rendering engine does not allow font files to be loaded at runtime via the fontconfig API.

You are very welcome to test the project.

**Comment by jcupitt** [comment]
Nice work @olivervogel! I had a quick look and made some suggestions.

**Comment by GuySartorelli** [comment]
That's awesome! Sorry I wasn't able to jump in and help like I thought I would.
